chapters/functions.tex

```diff
@@ -1567,13 +1567,21 @@ \section{Declaring Inverses of Functions}\label{declaring-inverses-of-functions}
 
 The function in the \lstinline!inverse! annotation must be an actual inverse.
 This requires that $f_2$ is such that, if $u_k$ is calculated as \lstinline!$u_k$ := $f_2$($\ldots$, y, $\ldots$)!, the equality \lstinline!y = $f_1$($\ldots$, $u_k$, $\ldots$)! is satisfied up to a certain precision, for all valid values of the input arguments of \lstinline!$f_2$($\ldots$, y, $\ldots$)!.
-the equality \lstinline!y = $f_1$($\ldots$, $u_k$, $\ldots$)! is satisfied up to a certain precision.
+
+\begin{nonnormative}
+There is no guarantee that a provided inverse will be used, and no rule for at which stage of symbolic processing it could be applied.
+Hence, providing an inverse function cannot be used as a means to select a particular solution to a nonlinear equation with multiple solutions (in other words, to select particular pre-image for a non-injective function).
+On the contrary, providing an inverse function can make a model ambiguous, as it will not be known if an equation is solved using a generic nonlinear equation solver (with solution being influenced by guess values), or solved using the provided inverse (possibly selecting a different solution).
+
+While it is possible to have several inverse functions that could be applied to solve the same equation, the real problem of ambiguity is not the number of applicable inverse functions, but that there is ambiguity as soon as there is one inverse function and multiple solutions.
```

Comment:
```suggestion
While it is possible to have several inverse functions that could be applied to solve the same equation, the real problem of ambiguity is not the number of applicable inverse functions, but that there is ambiguity as soon as there are multiple solutions.
Although an inverse function do not create this issue it can make it worse, in particular since it lacks access to the previous value to ensure continuity of the solution.
```

Reply:
In @HansOlsson's suggestion above the last sentence: "Although an inverse function do not..."  -> "... does not..."

Reply:
Updated!

Reply:
I think this change is worth discussing in some depth before applying, as it changes the original intention.

The main question is this: Do we really want to consider the existence of multiple solutions a problem on its own?  I mean, is any model having an equation with multiple solutions a problematic model?  I was hoping that we wouldn't have to be so hard on all these models – that Modelica is an environment where this can be considered normal, and that the combination of guess values and tracking of solutions over time are essential features that can make this possible.

Of course, it's not hard to come up with models where the existence multiple solutions (without any inverse functions) is really causing problems, for a number of different reasons.  But is the fact that some models have problems really reason to also consider the working ones problematic?

The point I was trying to make in the original formulation is that multiple solutions combined with an inverse function is where the real problem begins.  Since one can't control whether the inverse function will be used, the user will have to make sure that the same solution is obtained both when the inverse function is used and when it isn't.  To see what the solution is when it isn't used is tedious but easy – just remove the annotation – but making sure that the provided inverse function gives the same solution will require relatively advanced debugging of the translated model.

Maybe we won't have to bring this all the way back to a phone meeting poll, but considering @gkurzbach's [recent comment in a completely different issue](https://github.com/modelica/ModelicaSpecification/pull/2748#issuecomment-802814326), it would be interesting to at least hear from him.  @eshmoylova, should your comment on just the grammatical part be taken as being in favor of the change of intention?

Reply:
Reformulating an idea from the thread above, I also see a way out where changing the intention as proposed would make sense to me: To explicitly state that the effects of providing an inverse function are undefined when there are multiple solutions.  Then the introduction of inverse functions cannot make the situation worse, so the only problem that remains is the existence of multiple solutions.

(As we only use formulations with _undefined_ on rare occasions, it might be better to reformulate as a requirement that tools are not expected to verify.)

Reply:
> To explicitly state that the effects of providing an inverse function are undefined when there are multiple solutions. Then the introduction of inverse functions cannot make the situation worse, so the only problem that remains is the existence of multiple solutions.

Yes, the actual problem are multiple solution; and that was the point I was trying to make. How serious that problem is depends on the situation.

However, if we want to formulate it in other ways we have to be careful and not over-use undefined: many standards differentiate between undefined, implementation-defined and selecting among a few possible definitions as in this case.

Reply:
> However, if we want to formulate it in other ways we have to be careful and not over-use undefined: many standards differentiate between undefined, implementation-defined and selecting among a few possible definitions as in this case.

Yes.  At a glance, I can currently see only two uses of this phrasing: numeric overflow and impure functions in an algebraic loop.  While I find it hard to come up with something better for numeric overflow, I find the case of impure functions in an algebraic loop very similar to the current problem with inverse functions and multiple solutions.

Reply:
To me it depends, similarly as the _potential_ for numeric overflow.

Consider y=x^2 and giving the inverse sqrt(y). If all values were strictly above some positive number the fact that the function had multiple solution and the inverse picked one isn't a problem. (That could happen if the x is a temperature.)
But if x is negative or changes sign it's bad and using start-values etc will not always guarantee the correct solution.

That's why I don't think we should consider multiple solutions as very bad and undefined gives that feel; and also why I think that we shouldn't say that it's an issue with inverses - but with multiple solutions.

Reply:
You are pointing to a technical detail of my original proposal, which I excluded for brevity when repeating the idea: to avoid ending up in "undefined" territory – that is, to fulfill the injectiveness requirement before providing an inverse function – it is OK to take a informally specified domain of the function into consideration.

While possible to also express a function domain somewhat formally using asserts, there is little reason to require such formalization, since there is generally no practical way to verify the injectiveness property anyway.

A valid form of your example would thus be:
```
function squareNonNegative
  input Real x "Non-negative number";
  output Real y = x * x;
algorithm
  /* An assert can be used to make function domain somewhat explicit,
   * but is not required to specify an inverse function:
   */
//assert(x >= 0, "Invalid domain.");
  annotation(inverse(y = sqrt(x)));
end squareNonNegative;
```

Since it is completely unspecified how the function domain should be given, the following would also be a valid use of an inverse function as long as the function is only used on a domain where the inverse is correct:
```
function inverseOfSqrt
  input Real x;
  output Real y = x * x;
algorithm
  annotation(inverse(y = sqrt(x)));
end inverseOfSqrt;
```

Reply:
But I don't see how that prevents us from specifying that the underlying problem is attempting to invert non-injective functions, and that it works if the domain is in practice restricted.

Reply:
It's not an attempt at preventing us from saying that.  It's an attempt at saying that, but in a more formal manner: Instead of saying that validly providing inverse functions to non-injective functions can lead to trouble, we say that specifying inverse function isn't supported/legal/defined unless the function being inverted is effectively injective, considering an informally defined function domain when there is one.

Reply:
I still don't understand why my proposal above is controversial. The underlying issue is that we have a non-injective function and try to invert it; and we should present as the issue - not scare away users from providing inverses.

Reply:
I'm trying to avoid a situation where the specification should be blamed for allowing inverses to be defined in ill-defined ways.  You may be right that one needs to be careful not not scare people away from providing inverse functions when it is safe to do so, but I'm not at all against scaring users away from providing inverses in the ill-defined situations.

Reply:
My point is that the ill-defined situations are already problematic - because we have a non-injective functions that we for some reason try to invert. The inverses might make it a bit worse, but they don't create the problem.

Reply:
I think we have at least clearly identified the two different viewpoints, and isn't deciding between the two a very good topic for the phone meeting (hoping for enough attendance so that the language group can be considered informed and listened to)?

Reply:
Resolved according to phone meeting decision in https://github.com/modelica/ModelicaSpecification/pull/2901#issuecomment-831320374.